File: analytics/src/test/java/io/appmetrica/analytics/impl/startup/StartupUnitListenerNotificationTest.java
package io.appmetrica.analytics.impl.startup;

import io.appmetrica.analytics.coreutils.internal.time.TimeProvider;
import io.appmetrica.analytics.impl.ClidsInfoStorage;
import io.appmetrica.analytics.impl.GlobalServiceLocator;
import io.appmetrica.analytics.impl.clids.ClidsInfo;
import io.appmetrica.analytics.impl.component.ComponentId;
import io.appmetrica.analytics.impl.request.StartupRequestConfig;
import io.appmetrica.analytics.impl.startup.parsing.StartupResult;
import io.appmetrica.analytics.impl.startup.uuid.MultiProcessSafeUuidProvider;
import io.appmetrica.analytics.impl.startup.uuid.UuidValidator;
import io.appmetrica.analytics.impl.utils.DeviceIdGenerator;
import io.appmetrica.analytics.impl.utils.ServerTime;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.GlobalServiceLocatorRule;
import io.appmetrica.analytics.testutils.TestUtils;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class StartupUnitListenerNotificationTest extends CommonTest {

    @Rule
    public GlobalServiceLocatorRule globalServiceLocatorRule = new GlobalServiceLocatorRule();

    @Mock
    private StartupConfigurationHolder mStartupConfigurationHolder;
    @Mock
    private DeviceIdGenerator deviceIdGenerator;
    @Mock
    private StartupRequestConfig mStartupRequestConfig;
    @Mock
    private StartupResultListener mStartupResultListener;
    @Mock
    private ComponentId mComponentId;
    @Mock
    private StartupState.Storage mStartupStateStorage;
    @Mock
    private ClidsInfoStorage clidsStorage;
    @Mock
    private ClidsStateChecker clidsStateChecker;
    @Mock
    private MultiProcessSafeUuidProvider uuidProvider;
    @Mock
    private UuidValidator uuidValidator;
    private final String mPackageName = "test.package";
    private StartupUnit mStartupUnit;
    private StartupState mStartupState;

    @Before
    public void setup() {
        MockitoAnnotations.openMocks(this);
        when(mStartupConfigurationHolder.get()).thenReturn(mStartupRequestConfig);
        when(mComponentId.getPackage()).thenReturn(mPackageName);
    }

    @After
    public void tearDown() {
        GlobalServiceLocator.destroy();
    }

    @Test
    public void testGeneratedIdentifiersNotifies() {
        createStartupUnitToModifyStartup();
        checkListenersNotified(1);
    }

    @Test
    public void testDidNotGenerateIdentifiersButConstructorNotifies() {
        createStartupUnitToNotModifyStartup();
        checkListenersNotified(1);
    }

    @Test
    public void testOnRequestCompleteNotifies() {
        createStartupUnitToNotModifyStartup();
        checkListenersNotified(1);
        StartupResult startupResult = mock(StartupResult.class);
        ServerTime.getInstance().init();
        when(startupResult.getValidTimeDifference()).thenReturn(100L);
        when(startupResult.getCollectionFlags()).thenReturn(mock(CollectingFlags.class));
        when(mStartupRequestConfig.getChosenClids()).thenReturn(mock(ClidsInfo.Candidate.class));
        when(mStartupConfigurationHolder.getStartupState()).thenReturn(mStartupState);
        mStartupUnit.onRequestComplete(startupResult, mStartupRequestConfig, null);
        checkListenersNotified(2);
    }

    @Test
    public void testUpdateConfigurationNotifiesListener() {
        createStartupUnitToNotModifyStartup();
        checkListenersNotified(1);

        when(mStartupConfigurationHolder.getStartupState()).thenReturn(mStartupState);
        StartupRequestConfig.Arguments arguments = mock(StartupRequestConfig.Arguments.class);
        when(mStartupRequestConfig.getNewCustomHosts()).thenReturn(null);
        when(mStartupRequestConfig.getStartupHostsFromClient()).thenReturn(Arrays.asList("host"));
        when(mStartupRequestConfig.hasNewCustomHosts()).thenReturn(true);
        mStartupUnit.updateConfiguration(arguments);
        checkListenersNotified(2);
    }

    private void checkListenersNotified(int times) {
        List<Object> objects = new ArrayList<Object>();
        for (int i = 0; i < times; i++) {
            objects.add(mStartupConfigurationHolder);
            objects.add(GlobalServiceLocator.getInstance().getStartupStateHolder());
            objects.add(mStartupResultListener);
        }
        InOrder inOrder = Mockito.inOrder(objects.toArray());
        for (int i = 0; i < times; i++) {
            inOrder.verify(mStartupConfigurationHolder).updateStartupState(any(StartupState.class));
            inOrder.verify(GlobalServiceLocator.getInstance().getStartupStateHolder()).onStartupStateChanged(any(StartupState.class));
            inOrder.verify(mStartupResultListener).onStartupChanged(eq(mPackageName), any(StartupState.class));
        }
    }

    private void createStartupUnitToModifyStartup() {
        mStartupState = TestUtils.createDefaultStartupStateBuilder()
                .withUuid("uuid")
                .withDeviceId("deviceId")
                .build();
        when(uuidValidator.isValid("uuid")).thenReturn(true);
        when(mStartupConfigurationHolder.get()).thenReturn(mStartupRequestConfig);
        when(deviceIdGenerator.generateDeviceId()).thenReturn("deviceid");
        mStartupUnit = new StartupUnit(
                RuntimeEnvironment.getApplication(), mComponentId, mStartupResultListener, mStartupStateStorage,
                mStartupState, deviceIdGenerator, mStartupConfigurationHolder, mock(TimeProvider.class),
                clidsStorage, clidsStateChecker, uuidProvider, uuidValidator);
    }

    private void createStartupUnitToNotModifyStartup() {
        when(mStartupConfigurationHolder.get()).thenReturn(mStartupRequestConfig);
        mStartupState = TestUtils.createDefaultStartupStateBuilder()
                .withUuid("uuid")
                .withDeviceId("deviceId")
                .build();
        when(uuidValidator.isValid("uuid")).thenReturn(true);
        when(mStartupConfigurationHolder.get()).thenReturn(mStartupRequestConfig);
        mStartupUnit = new StartupUnit(
                RuntimeEnvironment.getApplication(), mComponentId, mStartupResultListener, mStartupStateStorage,
                mStartupState, deviceIdGenerator, mStartupConfigurationHolder, mock(TimeProvider.class),
                clidsStorage, clidsStateChecker, uuidProvider, uuidValidator);
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/startup/StartupUnitListenerNotificationTest.kt
package io.appmetrica.analytics.impl.startup

import io.appmetrica.analytics.coreapi.internal.identifiers.IdentifierStatus
import io.appmetrica.analytics.coreutils.internal.time.TimeProvider
import io.appmetrica.analytics.impl.ClidsInfoStorage
import io.appmetrica.analytics.impl.StartupStateHolder
import io.appmetrica.analytics.impl.clids.ClidsInfo
import io.appmetrica.analytics.impl.component.ComponentId
import io.appmetrica.analytics.impl.request.StartupRequestConfig
import io.appmetrica.analytics.impl.selfreporting.AppMetricaSelfReportFacade
import io.appmetrica.analytics.impl.selfreporting.SelfReporterWrapper
import io.appmetrica.analytics.impl.startup.parsing.StartupResult
import io.appmetrica.analytics.impl.startup.uuid.MultiProcessSafeUuidProvider
import io.appmetrica.analytics.impl.startup.uuid.UuidValidator
import io.appmetrica.analytics.impl.utils.DeviceIdGenerator
import io.appmetrica.analytics.impl.utils.ServerTime
import io.appmetrica.analytics.internal.IdentifiersResult
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.GlobalServiceLocatorRule
import io.appmetrica.analytics.testutils.TestUtils
import io.appmetrica.analytics.testutils.on
import io.appmetrica.analytics.testutils.staticRule
import org.junit.Rule
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.eq
import org.mockito.kotlin.inOrder
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

internal class StartupUnitListenerNotificationTest : CommonTest() {

    @get:Rule
    var globalServiceLocatorRule = GlobalServiceLocatorRule()

    private val selfReporter: SelfReporterWrapper = mock()

    @get:Rule
    val appMetricaSelfReportFacadeRule = staticRule<AppMetricaSelfReportFacade> {
        on { AppMetricaSelfReportFacade.getReporter() } doReturn selfReporter
    }

    private val packageName = "test.package"

    private val componentId: ComponentId = mock {
        on { `package` } doReturn packageName
    }

    private val deviceIdGenerator: DeviceIdGenerator = mock()
    private val startupResultListener: StartupResultListener = mock()

    private val startupRequestConfig: StartupRequestConfig = mock()

    private val startupConfigurationHolder: StartupConfigurationHolder = mock {
        on { get() } doReturn startupRequestConfig
    }

    private val startupStateStorage: StartupState.Storage = mock()
    private val clidsStorage: ClidsInfoStorage = mock()
    private val clidsStateChecker: ClidsStateChecker = mock()
    private val uuidProvider: MultiProcessSafeUuidProvider = mock {
        on { readUuid() } doReturn IdentifiersResult("uuid", IdentifierStatus.OK, null)
    }
    private val uuidValidator: UuidValidator = mock()
    private val timeProvider: TimeProvider = mock()
    private val startupStateHolder: StartupStateHolder = mock()
    private val requestConfigArguments: StartupRequestConfig.Arguments = mock()

    private val startupUnitComponents: StartupUnitComponents = mock {
        on { startupConfigurationHolder } doReturn startupConfigurationHolder
        on { deviceIdGenerator } doReturn deviceIdGenerator
        on { componentId } doReturn componentId
        on { resultListener } doReturn startupResultListener
        on { startupStateStorage } doReturn startupStateStorage
        on { clidsStorage } doReturn clidsStorage
        on { clidsStateChecker } doReturn clidsStateChecker
        on { multiProcessSafeUuidProvider } doReturn uuidProvider
        on { uuidValidator } doReturn uuidValidator
        on { timeProvider } doReturn timeProvider
        on { packageName } doReturn packageName
        on { startupStateHolder } doReturn startupStateHolder
        on { requestConfigArguments } doReturn requestConfigArguments
    }

    private lateinit var startupUnit: StartupUnit
    private lateinit var startupState: StartupState

    @Test
    fun generatedIdentifiersNotifies() {
        createStartupUnitToModifyStartup()
        checkListenersNotified(1)
    }

    @Test
    fun didNotGenerateIdentifiersButConstructorNotifies() {
        createStartupUnitToNotModifyStartup()
        checkListenersNotified(1)
    }

    @Test
    fun onRequestCompleteNotifies() {
        createStartupUnitToNotModifyStartup()
        checkListenersNotified(1)

        val collectingFlags = mock<CollectingFlags>()
        val chosenClids = mock<ClidsInfo.Candidate>()
        whenever(startupRequestConfig.chosenClids).thenReturn(chosenClids)
        val startupResult = mock<StartupResult> {
            on { validTimeDifference } doReturn 100L
            on { collectionFlags } doReturn collectingFlags
        }
        ServerTime.getInstance().init()
        whenever(startupConfigurationHolder.startupState).thenReturn(startupState)
        startupUnit.onRequestComplete(startupResult, startupRequestConfig, null)
        checkListenersNotified(2)
    }

    @Test
    fun updateConfigurationNotifiesListener() {
        createStartupUnitToNotModifyStartup()
        checkListenersNotified(1)
        whenever(startupConfigurationHolder.startupState).thenReturn(startupState)
        val arguments = mock<StartupRequestConfig.Arguments>()
        whenever(startupRequestConfig.newCustomHosts).thenReturn(null)
        whenever(startupRequestConfig.startupHostsFromClient).thenReturn(mutableListOf("host"))
        whenever(startupRequestConfig.hasNewCustomHosts()).thenReturn(true)
        startupUnit.updateConfiguration(arguments)
        checkListenersNotified(2)
    }

    private fun checkListenersNotified(times: Int) {
        val objects: MutableList<Any> = ArrayList()
        for (i in 0 until times) {
            objects.add(startupConfigurationHolder)
            objects.add(startupStateHolder)
            objects.add(startupResultListener)
        }
        inOrder(*objects.toTypedArray()) {
            for (i in 0 until times) {
                verify(startupConfigurationHolder).updateStartupState(any())
                verify(startupStateHolder).onStartupStateChanged(any())
                verify(startupResultListener).onStartupChanged(eq(packageName), any())
            }
        }
    }

    private fun createStartupUnitToModifyStartup() {
        startupState = TestUtils.createDefaultStartupStateBuilder()
            .withUuid("uuid")
            .withDeviceId("deviceId")
            .build()
        whenever(uuidValidator.isValid("uuid")).thenReturn(true)
        whenever(startupConfigurationHolder.get()).thenReturn(startupRequestConfig)
        whenever(startupConfigurationHolder.startupState).thenReturn(startupState)
        whenever(deviceIdGenerator.generateDeviceId()).thenReturn("deviceid")
        startupUnit = StartupUnit(startupUnitComponents)
        startupUnit.init()
    }

    private fun createStartupUnitToNotModifyStartup() {
        whenever(startupConfigurationHolder.get()).thenReturn(startupRequestConfig)
        startupState = TestUtils.createDefaultStartupStateBuilder()
            .withUuid("uuid")
            .withDeviceId("deviceId")
            .build()
        whenever(uuidValidator.isValid("uuid")).thenReturn(true)
        whenever(startupConfigurationHolder.get()).thenReturn(startupRequestConfig)
        whenever(startupConfigurationHolder.startupState).thenReturn(startupState)
        startupUnit = StartupUnit(startupUnitComponents)
        startupUnit.init()
    }
}
